allennlp/models/reading_comprehension/dialog_qa.py

```diff
@@ -0,0 +1,407 @@
+            # batch_size * max_qa_count, passage_length, word_embed_dim
+            p1_answer_marker = p1_answer_marker.view(batch_size * max_qa_count, passage_length)
+            p1_answer_marker_emb = self._prev_ans_marker(p1_answer_marker)
+            repeated_embedded_passage = torch.cat([repeated_embedded_passage, p1_answer_marker_emb], dim=-1)
+            if self._num_context_answers > 1:
+                p2_answer_marker = p2_answer_marker.view(batch_size * max_qa_count, passage_length)
+                p2_answer_marker_emb = self._prev_ans_marker(p2_answer_marker)
+                repeated_embedded_passage = torch.cat([repeated_embedded_passage, p2_answer_marker_emb], dim=-1)
+                if self._num_context_answers > 2:
+                    p3_answer_marker = p3_answer_marker.view(batch_size * max_qa_count, passage_length)
+                    p3_answer_marker_emb = self._prev_ans_marker(p3_answer_marker)
+                    repeated_embedded_passage = torch.cat([repeated_embedded_passage, p3_answer_marker_emb],
+                                                          dim=-1)
+            repeated_passage_lstm_mask = passage_lstm_mask.unsqueeze(1). \
+                repeat(1, max_qa_count, 1).view(batch_size * max_qa_count, passage_length)
+            repeated_encoded_passage = self._variational_dropout(self._phrase_layer(repeated_embedded_passage,
+                                                                        repeated_passage_lstm_mask))
+        else:
+            encoded_passage = self._variational_dropout(self._phrase_layer(embedded_passage, passage_lstm_mask))
+            repeated_encoded_passage = encoded_passage.unsqueeze(1).repeat(1, max_qa_count, 1, 1).view( \
+                batch_size * max_qa_count, passage_lstm_mask.size()[1], -1)
+
+        encoded_question = self._variational_dropout(self._phrase_layer(embedded_question, question_lstm_mask))
+        encoding_dim = encoded_question.size(-1)
+
+        # Shape: (batch_size * max_qa_count, passage_length, question_length)
+        passage_question_similarity = self._matrix_attention(repeated_encoded_passage, encoded_question)
+        # Shape: (batch_size * max_qa_count, passage_length, question_length)
+        passage_question_attention = util.last_dim_softmax(passage_question_similarity, question_mask)
+
+        # Shape: (batch_size * max_qa_count, passage_length, encoding_dim)
+        passage_question_vectors = util.weighted_sum(encoded_question, passage_question_attention)
+
+        # We replace masked values with something really negative here, so they don't affect the
+        # max below.
+        masked_similarity = util.replace_masked_values(passage_question_similarity,
+                                                       question_mask.unsqueeze(1),
+                                                       -1e7)
+
+        passage_mask = passage_mask.unsqueeze(1).repeat(1, max_qa_count, 1).view(batch_size * max_qa_count, -1)
```

Comment:
Instead of this line, I _think_ you can just do `util.last_dim_softmax(question_passage_similarity, passage_mask.unsqueeze(1))` below.